R/method-differentialAbundance.R

```diff
@@ -0,0 +1,194 @@
+    if (!'data.table' %in% class(df)) {
+      data.table::setDT(df)
+    }
+
+    # Check that incoming metadata meets requirements
+    if (!'data.table' %in% class(sampleMetadata)) {
+      data.table::setDT(sampleMetadata)
+    }
+
+    computeMessage <- ''
+    veupathUtils::logWithTime(paste("Received df table with", nrow(df), "samples and", (ncol(df)-1), "taxa."), verbose)
+
+    if (naToZero) {
+      # Replace NA values with 0
+      veupathUtils::setNaToZero(df)
+      veupathUtils::logWithTime("Replaced NAs with 0", verbose)
+    }
+
+    ## Check that groups are provided, if needed, and if they are provided,
+    ## that they match at least one value in the comparisonVariable column.
+    uniqueComparisonVariableValues <- sort(unique(sampleMetadata[[comparisonVariable]]))
+    
+    if (!!length(groupA) && !!length(groupB)) {
+      # Does each group contain at least one string that matches a value in the comparisonValue column?
+      if (purrr::none(groupA,function(x, arr) { x %in% arr}, uniqueComparisonVariableValues)) {
+        stop("At least one value in groupA must exist as a value in the comparisonValue sampleMetadata column.")
+      }
+      if (purrr::none(groupB,function(x, arr) { x %in% arr}, uniqueComparisonVariableValues)) {
+        stop("At least one value in groupB must exist as a value in the comparisonValue sampleMetadata column.")
+      }
+
+      # Alert that we discard groupA/B values that are not found in the comparisonVariable
+      if (any(!(groupA %in% uniqueComparisonVariableValues))) {
+        veupathUtils::logWithTime("Found values in groupA that do not exist in the comparisonVariable. Removing these values.", verbose)
+        groupA <- groupA[groupA %in% uniqueComparisonVariableValues]
+      }
+      if (any(!(groupB %in% uniqueComparisonVariableValues))) {
+        veupathUtils::logWithTime("Found values in groupB that do not exist in the comparisonVariable. Removing these values.", verbose)
+        groupB <- groupB[groupB %in% uniqueComparisonVariableValues]
+      }
+
+      # Do not allow duplicated values
+      if (!!length(intersect(groupA, groupB))) {
+        veupathUtils::logWithTime("groupA and groupB cannot share members.", verbose)
+        stop()
+      }
+
+    } else if (length(uniqueComparisonVariableValues) == 2) {
+      # Ignore any groups passed to us and set the groups to be these two values
+      veupathUtils::logWithTime("Only two values found in the comparisonVariable sampleMetadata column. Using these two values as groupA and groupB.", verbose)
+      groupA <- c(uniqueComparisonVariableValues[1])
+      groupB <- c(uniqueComparisonVariableValues[2])
+    } else {
+      stop("Must supply two groups (groupA and groupB) for the differential abundance calculation.")
+    }
+
+    # Subset to only include samples with metadata defined in groupA and groupB
+    sampleMetadata <- sampleMetadata[get(comparisonVariable) %in% c(groupA, groupB), ]
```

Comment:
Yes that issue is [here](https://github.com/VEuPathDB/web-monorepo/issues/17), and i should get that discussion going quickly since i think it'll be confusing to see far fewer points than one expects.